fast64_internal/f3d/f3d_material.py

```diff
@@ -177,85 +177,144 @@ def update_draw_layer(self, context):
-                and f3dMat.rdp_settings.blend_a2 == "G_BL_A_IN"
-                and f3dMat.rdp_settings.blend_m2 == "G_BL_CLR_MEM"
-                and f3dMat.rdp_settings.blend_b2 == "G_BL_1MA"
-            ):
-                blend_method = "BLEND"
-            else:
-                blend_method = "OPAQUE"
-        else:
-            rendermode = f3dMat.rdp_settings.rendermode_preset_cycle_1
-            if not is_one_cycle:
-                rendermode = f3dMat.rdp_settings.rendermode_preset_cycle_2
-
-            f3d = get_F3D_GBI()
-            r_mode = getattr(f3d, rendermode, f3d.G_RM_AA_ZB_OPA_SURF)
-            if r_mode & f3d.CVG_X_ALPHA:
-                blend_method = "CLIP"
-            else:
-                cfunc = GBL_c1 if is_one_cycle else GBL_c2
-                xlu_comb = r_mode & cfunc(f3d.G_BL_CLR_IN, f3d.G_BL_A_IN, f3d.G_BL_CLR_MEM, f3d.G_BL_1MA)
-                if xlu_comb and r_mode & f3d.FORCE_BL:
-                    blend_method = "BLEND"
-                else:
-                    blend_method = "OPAQUE"
 
-    return blend_method
+def does_blender_use_mix(
+    settings: "RDPSettings",
+    mix: str,
+    default_for_no_rendermode: bool = False
+) -> bool:
+    if not settings.set_rendermode:
+        return default_for_no_rendermode
+    is_one_cycle = settings.g_mdsft_cycletype == "G_CYC_1CYCLE"
+    return settings.blend_b1 == mix or (is_one_cycle and settings.blend_b2 == mix)
+
+
+def is_blender_lerp(
+    settings: "RDPSettings",
+    cycle: int,
+    p: str,
+    a: str,
+    m: str,
+    b: str,
+    default_for_no_rendermode: bool = False
+) -> bool:
+    assert cycle in {1, 2, -1}  # -1 = last cycle
+    if cycle == -1:
+        cycle = 1 if settings.g_mdsft_cycletype == "G_CYC_1CYCLE" else 2
+    if not settings.set_rendermode:
+        return default_for_no_rendermode
+    return (
+        getattr(settings, f"blend_p{cycle}") == p
+        and getattr(settings, f"blend_a{cycle}") == a
+        and getattr(settings, f"blend_m{cycle}") == m
+        and getattr(settings, f"blend_b{cycle}") == b
+    )
+
+
+def is_blender_doing_fog(settings: "RDPSettings") -> bool:
+    return is_blender_lerp(
+        settings,
+        # If 2 cycle, fog must be in first cycle.
+        1,
+        "G_BL_CLR_FOG",
+        "G_BL_A_SHADE",
+        # While technically it being fog only requires that P and A be fog color
+        # and shade alpha, the only reasonable choice for M and B in this case
+        # is color in and 1-A.
+        "G_BL_CLR_IN",
+        "G_BL_1MA",
+        # if NOT setting rendermode, it is more likely that the user is setting
+        # rendermodes in code, so to be safe we'll enable fog
+        True
+    )
+
+
+def get_blend_method(material: bpy.types.Material) -> str:
+    settings = material.f3d_mat.rdp_settings
+    if not settings.set_rendermode:
+        return drawLayerSM64Alpha[material.f3d_mat.draw_layer.sm64]
+    if settings.cvg_x_alpha:
+        return "CLIP"
```

Comment:
I have yet to entirely figure out the rdp state about alpha clip and blend (i.e. am not an expert), but I'd expect `g_mdsft_alpha_compare` to play a role more than cvg_x_alpha?

(more of a open-ended rdp question sorry)

Reply:
This is not new code by me--this is the function `get_blend_method` originally at line 210. It has been refactored because it no longer needs to check all the details of the rendermode, instead directly using the RDP settings and a helper function.